services/gitlab/gitlab-pipeline-status.service.js

```diff
@@ -0,0 +1,108 @@
+'use strict'
+
+const Joi = require('joi')
+const validate = require('../../lib/validate')
+const BaseSvgScrapingService = require('../base-svg-scraping')
+const { InvalidParameter, NotFound } = require('../errors')
+const { isPipelineStatus } = require('./gitlab-helpers')
+
+const badgeSchema = Joi.object({
+  message: Joi.alternatives()
+    .try([isPipelineStatus, Joi.equal('unknown')])
+    .required(),
+}).required()
+
+const queryParamsSchema = Joi.object({
+  // TODO This accepts URLs with query strings and fragments, which should be
+  // rejected.
+  gitlab_url: Joi.string().uri({ scheme: ['https'] }),
```

Comment:
In this case, maybe it would be easier to validate and then work out if we've got a queryString (or strip it away) by parsing into a `URLObject` instead of using Joi:

```js
> const { URL } = require('url')
> const url = new URL('http://foo.bar/baz?foo=bar')
> url
URL {
  href: 'http://foo.bar/baz?foo=bar',
  origin: 'http://foo.bar',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'foo.bar',
  hostname: 'foo.bar',
  port: '',
  pathname: '/baz',
  search: '?foo=bar',
  searchParams: URLSearchParams { 'foo' => 'bar' },
  hash: '' }
```

Reply:
That could work… though it's a somewhat messy thing to replicate across other services that might use this. Adding some flags to the Joi function seems like a better way to handle it.

Checking for valid URLs is a good start. It doesn't feel important to get this perfect.